test/integration/test_uv_truncate_snapshot.c

```diff
@@ -0,0 +1,247 @@
+            raft_free(_snap->bufs[0].base);                               \
+            raft_free(_snap->bufs);                                       \
+            raft_free(_snap);                                             \
+        }                                                                 \
+    } while (0);
+
+#define SNAPSHOT_PUT_REQ(TRAILING, INDEX, RV, STATUS)              \
+    struct raft_snapshot _snapshot;                                \
+    struct raft_buffer _snapshot_buf;                              \
+    uint64_t _snapshot_data;                                       \
+    struct raft_io_snapshot_put _req;                              \
+    struct result _result = {STATUS, false, NULL};                 \
+    int _rv;                                                       \
+    _snapshot.term = 1;                                            \
+    _snapshot.index = INDEX;                                       \
+    raft_configuration_init(&_snapshot.configuration);             \
+    _rv = raft_configuration_add(&_snapshot.configuration, 1, "1", \
+                                 RAFT_STANDBY);                    \
+    munit_assert_int(_rv, ==, 0);                                  \
+    _snapshot.bufs = &_snapshot_buf;                               \
+    _snapshot.n_bufs = 1;                                          \
+    _snapshot_buf.base = &_snapshot_data;                          \
+    _snapshot_buf.len = sizeof _snapshot_data;                     \
+    _req.data = &_result;                                          \
+    _rv = f->io.snapshot_put(&f->io, TRAILING, &_req, &_snapshot,  \
+                             snapshotPutCbAssertResult);           \
+    munit_assert_int(_rv, ==, RV)
+
+#define SNAPSHOT_CLEANUP() raft_configuration_close(&_snapshot.configuration)
+
+/******************************************************************************
+ *
+ * test interaction of raft_io->snapshot_put and raft_io->truncate()
+ *
+ *****************************************************************************/
+
+SUITE(snapshot_truncate)
+
+/* Fill up 3 segments worth of data, then take a snapshot.
+ * While the snapshot is taken, start a truncate request. */
+TEST(snapshot_truncate, snapshotThenTruncate, setUp, tearDownDeps, 0, NULL)
+{
+    struct fixture *f = data;
+    APPEND_SUBMIT(0, MAX_SEGMENT_BLOCKS, SEGMENT_BLOCK_SIZE);
+    APPEND_SUBMIT(1, MAX_SEGMENT_BLOCKS, SEGMENT_BLOCK_SIZE);
+    APPEND_SUBMIT(2, MAX_SEGMENT_BLOCKS, SEGMENT_BLOCK_SIZE);
+
+    /* Take a snapshot, this will use a uv_barrier. */
+    SNAPSHOT_PUT_REQ(8192, 6, 0, 0);
+
+    /* Truncate, this will use a uv_barrier too.  */
+    TRUNCATE(8);
+
+    /* There's no truncate callback to wait for, loop for a while. */
+    LOOP_RUN(1000);
+
+    /* Check that truncate has done its job. */
+    ASSERT_ENTRIES(7, 1, 2, 3, 4, 5, 6, 7);
+
+    SNAPSHOT_CLEANUP();
+    return MUNIT_OK;
+}
```

Comment:
Still trying to understand exactly the situation: if we managed to drop non-blocking barriers (by having a way to differentiate between open-segment-starting-at-1 and open-segment-past-a-snapshot), then there should be no need to have a barrier when a snapshot is taken right? I'm wondering if that'd mean that this test would pass.

In other words, the "run a non-blocking barrier when a snapshot is taken" change fixed the differentiation problem, but introduced the bug being fixed here. Is that accurate?

Reply:
> Indeed in the past Barriers were only used for Truncate operations, that are more rare than snapshots. With multiple truncate requests it's probably possible to hit the bug though (that was always there), but it is a lot easier to hit now.
> 
> An issue I encountered when removing the barrier when taking a snapshot, was that the truncate operation that runs after the snapshot (to remove a prefix of the log) could remove all closed segments, again leading to confusion if the open segments remaining in the data folder were newer or older than the snapshot. In the case of taking a snapshot and all closed segments being removed, the open segments were older, in the case of installing a snapshot the open segments will be newer, with no meaningful way to discern the situation. I then more or less felt that the barrier before a snapshot made the situation easier to reason about and that it was not that bad in the end (imo :-))
> 
> We could have come up with special rules when taking a snapshot like "Don't remove the last closed segment when truncating a prefix of the log otherwise we will be confused", but that feels very ad hoc and the meaning of the `Trailing` parameter loses its value somewhat. Or we could start adding special files in the directory to discern the situations but that also quickly becomes a mess IMO. The right solution would be to create a new segment format that encodes the start index, but it's a big(ish) change and we break backwards compatibility.

Ok, thanks for the write up, I can better see the whole picture now.

If there's something that simplifies the situation, that'd be good, otherwise ad-hoc rules are just going to flip the problem around.

Just for my understanding: a new segment forward would break forward compatibility, right? In the sense that it'd be trick to support downgrades from a version that supports the new format to a version that doesn't. However backward compatibility should be fine (upgrading from old version to new version). Is that accurate?

Reply:
Yes that's accurate, got confused by the terminology.

Reply:
Ok, what about merging this change and at the same time opening a (longer term) issue to investigate the format change or other ideas?

If and when we find a solution that avoids the downsides you described, we could then simplify the code by removing non-blocking barriers and assume that barriers are used only for truncation (that assumption should help too for reducing complexity).

Reply:
https://github.com/canonical/raft/issues/453

Reply:
> #453

Nice write-up, thanks.